go/libraries/doltcore/merge/revert.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package merge
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dolthub/dolt/go/cmd/dolt/cli"
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+)
+
+func Revert(ctx context.Context, ddb *doltdb.DoltDB, workingRoot *doltdb.RootValue, commits []*doltdb.Commit) (*doltdb.RootValue, string, error) {
+	revertMessage := "Revert"
+
+	for i, baseCommit := range commits {
+		if i > 0 {
+			revertMessage += " and"
+		}
+		baseRoot, err := baseCommit.GetRootValue()
+		if err != nil {
+			return nil, "", err
+		}
+		baseMeta, err := baseCommit.GetCommitMeta()
+		if err != nil {
+			return nil, "", err
+		}
+		revertMessage = fmt.Sprintf(`%s "%s"`, revertMessage, baseMeta.Description)
+
+		var theirRoot *doltdb.RootValue
+		if len(baseCommit.ParentRefs()) > 0 {
+			parentCM, err := ddb.ResolveParent(ctx, baseCommit, 0)
+			if err != nil {
+				return nil, "", err
+			}
+			theirRoot, err = parentCM.GetRootValue()
+			if err != nil {
+				return nil, "", err
+			}
+		} else {
+			theirRoot, err = doltdb.EmptyRootValue(ctx, ddb.ValueReadWriter())
+			if err != nil {
+				return nil, "", err
+			}
+		}
+
+		workingRoot, _, err = MergeRoots(ctx, workingRoot, theirRoot, baseRoot)
+		if err != nil {
+			return nil, "", err
+		}
+		if ok, err := workingRoot.HasConflicts(ctx); err != nil {
+			return nil, "", err
+		} else if ok {
+			cli.PrintErrln("revert currently does not handle conflicts")
```

Comment:
these only work for CLI error reporting

Reply:
Oh whoops, I need to be returning these as regular errors, well you can see where I wrote the original draft lol